types/utils.go

```diff
@@ -21,58 +22,85 @@ var (
 	errUnsupportedKeyType = errors.New("unsupported key type")
 )
 
-// GetRandomValidatorSet returns a validator set with a single validator
-func GetRandomValidatorSet() *cmtypes.ValidatorSet {
-	valSet, _ := GetRandomValidatorSetWithPrivKey()
-	return valSet
+type ValidatorConfig struct {
+	PrivKey     ed25519.PrivKey
+	VotingPower int64
 }
 
-// GetRandomValidatorSetWithPrivKey returns a validator set with a single
-// validator
-func GetRandomValidatorSetWithPrivKey() (*cmtypes.ValidatorSet, ed25519.PrivKey) {
-	privKey := ed25519.GenPrivKey()
-	pubKey := privKey.PubKey()
-	valset := cmtypes.NewValidatorSet(
-		[]*cmtypes.Validator{
-			{PubKey: pubKey, Address: pubKey.Address(), VotingPower: 1},
-		},
-	)
-	return valset, privKey
+// GetValidatorSet creates a validatorConfig with a randomly generated privateKey and votingPower set to 1,
+// then calls GetValidatorSetCustom to return a new validator set along with the validatorConfig.
+func GetValidatorSet() (*cmtypes.ValidatorSet, ValidatorConfig) {
+	config := ValidatorConfig{
+		PrivKey:     ed25519.GenPrivKey(),
+		VotingPower: 1,
+	}
+	valset := GetValidatorSetCustom(config)
+	return valset, config
 }
 
-// GetValidatorSet returns a validator set with a single validator
-// with the given key
-func GetValidatorSet(privKey ed25519.PrivKey) *cmtypes.ValidatorSet {
-	pubKey := privKey.PubKey()
+// GetValidatorSetCustom returns a validator set based on the provided validatorConfig.
+func GetValidatorSetCustom(config ValidatorConfig) *cmtypes.ValidatorSet {
+	if config.PrivKey == nil {
+		panic(errors.New("private key is nil"))
+	}
+	pubKey := config.PrivKey.PubKey()
 	valset := cmtypes.NewValidatorSet(
 		[]*cmtypes.Validator{
-			{PubKey: pubKey, Address: pubKey.Address(), VotingPower: 1},
+			{PubKey: pubKey, Address: pubKey.Address(), VotingPower: config.VotingPower},
 		},
 	)
 	return valset
 }
 
-// GetRandomBlock returns a block with random data
+// BlockContext carries all necessary state for block generation, including configurations and outputs.
+type BlockConfig struct {
+	Height       uint64
+	NTxs         int
+	PrivKey      ed25519.PrivKey // Input and Output option
+	ProposerAddr []byte          // Input option
+	Block        *Block          // Output: the generated block.
```

Comment:
I really don't see why one structure is used to both input and output. The typical convention in Go is just to return value.
The code is not simplified. What is the benefit?

Reply:
I believe I should change this to return the Block and the private key.
I would return the block, but sometimes, the private key is also expected in the return value.
But you are right, having a common struct is not a clean solution 😞
I'll push a better(hopefully) solution asap!